File: elixir/bob/bob.exs
defmodule Bob do

  def hey(message) do
    cond do
      Language.is_silence?(message)   -> "Fine. Be that way!"
      Language.is_question?(message)  -> "Sure."
      Language.is_aggresive?(message) -> "Whoa, chill out!"
      true                            -> "Whatever."
    end
  end

end

defmodule Language do

  def is_aggresive?(message) do
    String.upcase(message) == message &&
      String.downcase(message) != message
  end

  def is_question?(message) do
    String.ends_with?(message, "?")
  end

  def is_silence?("") do
    true
  end

  def is_silence?(message) do
    false
  end

end

File: haskell/bob/Bob.hs
module Bob (responseFor) where
import Data.Char

data EmotionalContent = Silence | Aggression | Question | Meh

responseFor :: String -> String
responseFor = display . classify

display :: EmotionalContent -> String
display Silence    = "Fine. Be that way!"
display Aggression = "Woah, chill out!"
display Question   = "Sure."
display Meh        = "Whatever."

classify :: String -> EmotionalContent
classify message
  | isSilence    message = Silence
  | isAggressive message = Aggression
  | isQuestion   message = Question
  | otherwise            = Meh

isAggressive :: String -> Bool
isAggressive message = any isUpper message && all (not . isLower) message

isQuestion :: String -> Bool
isQuestion message = not (null message) && last message == '?'

isSilence :: String -> Bool
isSilence = all isSpace
